omega/omega-transaction/src/main/java/org/apache/servicecomb/pack/omega/transaction/AbstractRecoveryPolicy.java

```diff
@@ -32,11 +32,20 @@ public abstract Object applyTo(ProceedingJoinPoint joinPoint, Compensable compen
   public Object apply(ProceedingJoinPoint joinPoint, Compensable compensable,
       CompensableInterceptor interceptor, OmegaContext context, String parentTxId, int retries)
       throws Throwable {
+    Object result;
     if(compensable.timeout()>0){
       RecoveryPolicyTimeoutWrapper wrapper = new RecoveryPolicyTimeoutWrapper(this);
-      return wrapper.applyTo(joinPoint, compensable, interceptor, context, parentTxId, retries);
+      result = wrapper.applyTo(joinPoint, compensable, interceptor, context, parentTxId, retries);
     }else{
-      return this.applyTo(joinPoint, compensable, interceptor, context, parentTxId, retries);
+      result = this.applyTo(joinPoint, compensable, interceptor, context, parentTxId, retries);
     }
+    if (compensable.sendingSagaEnd()) {
+      // Just send out the SagaEnd event
+      // TODO we may also invoke the callback here to release some resources
+      interceptor.sendSagaEndEvent();
```

Comment:
If there are multiple services that need to be compensated, this will cause Alpha to receive duplicate SagaEndedEvents.

Reply:
Alpha will receive the following events?

SagaStartedEvent
TxStartedEvent-1
TxEndedEvent-1
TxStartedEvent-2
TxEndedEvent-2
TxStartedEvent-3
TxAbortedEvent-3
SagaAbortedEvent
TxCompensabledEvent-2
SagaEndedEvent
TxCompensabledEvent-1
SagaEndedEvent

Reply:
SagaEnd just close the Saga transaction, can we just keep it open and let Alpha close after the timeout?
If there is a async invocation, we can send the async transaction start event first to Alpha,  then we can let the Alpha track the whole states to tell if the Saga transaction is OK to close.

Reply:
But the problem is how the alpha know the transaction is OK to close ? In the async situation, the alpha has no knowledge with the number of the participates, since the TxStartedEvent-n might be always on the road. The timeout is not a good idea as the saga is typically long running transaction.

Reply:
> SagaEnd just close the Saga transaction, can we just keep it open and let Alpha close after the timeout?

I don't recommend it, this introduces new uncertainty

> If there is an async invocation, we can send the async transaction start event first to Alpha, then we can let the Alpha track the whole states to tell if the Saga transaction is OK to close.

We define an attribute in the SagaStart annotation to set the number of sub-transactions for this global transaction. So we don't have to care about async problems, Alpha can use internal counters to judge that all sub-transactions have ended.

Reply:
Don't like sub-transaction number, that makes user side arch hard to change and cause bugs.
BTW, I think we can discuss this issue in the mail list.